### ISSUE TITLE
Typescript: emit compile error for unknown design tokens

### ISSUE THREAD

**Issue by anilanar** [chakra-ui/chakra-ui#4810]
### Description

Currently, there's no way to enforce type-safety for design tokens. I can just write `<Box backgroundColor="anything" />` and I have no clue that there's no color called `anything`. I have to catch this mistake in automated visual regression tests. I would rather catch them before that.

### Problem Statement/Justification

Right now, it is very likely to introduce accidental bugs with very hard to catch typos. The whole point of using typescript is to protect ourselves from such silly mistakes.

I respect everybody else who does not need this safety net, but I'm pretty sure there are others who would prefer this.

### Proposed Solution or API

We can aim for couple different options:

## 1. An object containing all tokens

```js
<Box color={t.colors['red-500']} />
```

Obviously, this is not ideal but probably better than nothing.

## 2. Some kind of smart type inference

```js
<Box color={t('red-500')} backgroundColor="#fff" />
```

Basically this would allow both typesafe tokens and also free-style values.

`type t = <ThemeKey>(token: keyof ThemeTypings[ThemeKey]) => TypesafeToken<ThemeKey>` and something like `type TypesafeToken<ThemeKey> = { __internal: 'TypesafeToken', key: ThemeKey, token: keyof ThemeTypings[ThemeKey] }`. These types don't have to really represent the reality, so `t` can be just an identity function `const t = x => x`. The problem here is how to make TS infer `ThemeKey` generic param correctly. For that to work, `color` property needs to have a `TypesafeToken<'Colors'>` type in its union. Even then, not sure if this is possible so have to try a proof of concept.
